simulations/vip-129/vip-129/simulations.ts

```diff
@@ -0,0 +1,152 @@
+import { expect } from "chai";
+import { parseUnits } from "ethers/lib/utils";
+import { ethers } from "hardhat";
+
+import { expectEvents, setMaxStalePeriodInChainlinkOracle } from "../../../src/utils";
+import { forking, testVip } from "../../../src/vip-framework";
+import { Actions, vip129 } from "../../../vips/vip-129/vip-129";
+import TUSD_ABI from "./abi/IERC20UpgradableAbi.json";
+import VTUSD_ABI from "./abi/VBep20Abi.json";
+import VBEP20_DELEGATOR_ABI from "./abi/VBep20DelegatorAbi.json";
+import COMPTROLLER_ABI from "./abi/comptroller.json";
+import PRICE_ORACLE_ABI from "./abi/resilientOracle.json";
+
+const COMPTROLLER = "0xfd36e2c2a6789db23113685031d7f16329158384";
+const NEW_VTUSD = "0xBf762cd5991cA1DCdDaC9ae5C638F5B5Dc3Bee6E";
+const OLD_VTUSD = "0x08CEB3F4a7ed3500cA0982bcd0FC7816688084c3";
+const VTOKEN_IMPLEMENTATION = "0x13f816511384D3534783241ddb5751c4b7a7e148"; // Original implementation
+const NEW_TUSD = "0x40af3827F39D0EAcBF4A168f8D4ee67c121D11c9";
+const OLD_TUSD = "0x14016E85a25aeb13065688cAFB43044C2ef86784";
+const CHAINLINK_ORACLE = "0x1B2103441A0A108daD8848D8F5d790e4D402921F";
+const ORACLE_FEED = "0xa3334A9762090E827413A7495AfeCE76F41dFc06";
+const NORMAL_TIMELOCK = "0x939bD8d64c0A9583A7Dcea9933f7b21697ab6396";
+const VTOKEN_RECEIVER = "0xc444949e0054A23c44Fc45789738bdF64aed2391";
+
+forking(29178963, () => {
+  let comptroller: ethers.Contract;
+  let tusd: ethers.Contract;
+  let vTusdOld: ethers.Contract;
+  let vTusd: ethers.Contract;
+  let oracle: ethers.Contract;
+  const provider = ethers.provider;
+
+  before(async () => {
+    comptroller = new ethers.Contract(COMPTROLLER, COMPTROLLER_ABI, provider);
+    oracle = new ethers.Contract(CHAINLINK_ORACLE, PRICE_ORACLE_ABI, provider);
+    tusd = new ethers.Contract(NEW_TUSD, TUSD_ABI, provider);
+    vTusdOld = new ethers.Contract(OLD_VTUSD, VTUSD_ABI, provider);
+    vTusd = new ethers.Contract(NEW_VTUSD, VTUSD_ABI, provider);
+    await setMaxStalePeriodInChainlinkOracle(CHAINLINK_ORACLE, OLD_TUSD, ORACLE_FEED, NORMAL_TIMELOCK);
+  });
+
+  testVip("VIP-129 TUSD Contract Migration", vip129(), {
+    callbackAfterExecution: async txResponse => {
+      await expectEvents(
+        txResponse,
+        [COMPTROLLER_ABI, VBEP20_DELEGATOR_ABI],
+        [
+          "ActionPausedMarket",
+          "NewImplementation",
+          "MarketListed",
+          "NewSupplyCap",
+          "NewBorrowCap",
+          "VenusSupplySpeedUpdated",
+          "VenusBorrowSpeedUpdated",
+          "NewCollateralFactor",
+          "NewReserveFactor",
+        ],
+        [3, 2, 1, 1, 1, 2, 2, 2, 1],
+      );
+    },
+  });
+
+  describe("Post-VIP behavior", async () => {
```

Comment:
Nit: I would also expect a check for CF of the old market, but I guess it works since `NewCollateralFactor` event is checked above. `_setCollateralFactor` doesn't revert on failures in the core pool AFAIR, so errors might be hard to catch.

Reply:
I added the missing checks in [5f10a63](https://github.com/VenusProtocol/vips/pull/25/commits/5f10a63e2955ba424462bc80cc620ea7c5b197d0)